sw/device/tests/penetrationtests/firmware/sca/aes_sca.c

```diff
@@ -630,6 +643,409 @@ status_t handle_aes_sca_fvsr_key_start_batch_generate(ujson_t *uj) {
+  }
+
+  // Trigger the AES GCM operation.
+  dif_aes_data_t aes_ctx_acc;
+  dif_aes_data_t aes_tag_acc;
+  memset(aes_ctx_acc.data, 0,
+         sizeof(aes_ctx_acc.data[0] * ARRAYSIZE(aes_ctx_acc.data)));
+  memset(aes_tag_acc.data, 0,
+         sizeof(aes_tag_acc.data[0] * ARRAYSIZE(aes_tag_acc.data)));
+  for (size_t it = 0; it < uj_num_ops.num_batch_ops; it++) {
+    dif_aes_data_t aes_ctx;
+    dif_aes_data_t aes_tag;
+    TRY(trigger_aes_gcm(
+        key_fvsr[it], aes_iv_fvsr[it], aes_aad, uj_aad_blocks.num_blocks,
+        aes_aad_last_block_size, aes_ptx, uj_ptx_blocks.num_blocks,
+        aes_ptx_last_block_size, &aes_ctx, &aes_tag, uj_triggers));
+    // Accumulate (i.e., XOR) TAG and CTX for sending back to host.
+    for (size_t i = 0; i < ARRAYSIZE(aes_ctx_acc.data); i++) {
+      aes_ctx_acc.data[i] ^= aes_ctx.data[i];
+      aes_tag_acc.data[i] ^= aes_tag.data[i];
+    }
+  }
+
+  // Send accumulated CTX and TAG back to host.
+  aes_sca_block_t uj_ctx;
+  aes_sca_block_t uj_tag;
+
+  uj_ctx.num_valid_bytes = 16;
+  memset(uj_ctx.block, 0, sizeof(uj_ctx.block));
+  memcpy(uj_ctx.block, (uint8_t *)aes_ctx_acc.data, uj_ctx.num_valid_bytes);
+
+  uj_tag.num_valid_bytes = 16;
+  memset(uj_tag.block, 0, sizeof(uj_tag.block));
+  memcpy(uj_tag.block, (uint8_t *)aes_tag_acc.data, uj_tag.num_valid_bytes);
+
+  RESP_OK(ujson_serialize_aes_sca_block_t, uj, &uj_ctx);
+  RESP_OK(ujson_serialize_aes_sca_block_t, uj, &uj_tag);
+
+  return OK_STATUS();
+}
+
+status_t handle_aes_sca_gcm_single_encrypt(ujson_t *uj) {
+  // Receive the AES-GCM input data over uJSON.
+  aes_sca_gcm_triggers_t uj_triggers;
+  aes_sca_block_t uj_iv;
+  aes_sca_key_t uj_key;
+  aes_sca_num_blocks_t uj_aad_blocks;
+  aes_sca_num_blocks_t uj_ptx_blocks;
+  aes_sca_block_t uj_aad[kMaxGcmBlocks];
+  aes_sca_block_t uj_ptx[kMaxGcmBlocks];
+  // Get the trigger configuration.
+  // uj_triggers.triggers[0] = True/False - process initial counter block.
+  // uj_triggers.triggers[1] = True/False - process AAD blocks.
+  // uj_triggers.triggers[2] = True/False - process PTX blocks.
+  // uj_triggers.block = int - which AAD or PTX block is captured?
+  TRY(ujson_deserialize_aes_sca_gcm_triggers_t(uj, &uj_triggers));
+  // Get IV and KEY.
+  TRY(ujson_deserialize_aes_sca_block_t(uj, &uj_iv));
+  TRY(ujson_deserialize_aes_sca_key_t(uj, &uj_key));
+  // Get number of AAD and PTX blocks we are expecting.
+  TRY(ujson_deserialize_aes_sca_num_blocks_t(uj, &uj_aad_blocks));
+  TRY(ujson_deserialize_aes_sca_num_blocks_t(uj, &uj_ptx_blocks));
+  if (uj_aad_blocks.num_blocks > kMaxGcmBlocks ||
+      uj_ptx_blocks.num_blocks > kMaxGcmBlocks) {
+    return ABORTED();
+  }
+  // Fetch all AAD blocks.
+  for (size_t it = 0; it < uj_aad_blocks.num_blocks; it++) {
+    TRY(ujson_deserialize_aes_sca_block_t(uj, &uj_aad[it]));
+  }
+  // Fetch all PTX blocks.
+  for (size_t it = 0; it < uj_ptx_blocks.num_blocks; it++) {
+    TRY(ujson_deserialize_aes_sca_block_t(uj, &uj_ptx[it]));
+  }
```

Comment:
I was first surprised to see that we're getting all AAD and PTX blocks but it makes sense as we keep those fixed: we actually want to analyze the key and IV.